src/laddie/laddie_velocity.f90

```diff
@@ -0,0 +1,420 @@
+          u_perp_x = U_c_tot( ei) * D_x/D_c
+          u_perp_y = V_c_tot( ei) * D_y/D_c
+
+          ! Calculate upstream momentum divergence
+          ! =============================
+          ! u_perp > 0: flow is exiting this triangle into triangle tj
+          IF (u_perp_x > 0) THEN
+            laddie%divQU( ti) = laddie%divQU( ti) + mesh%TriCw( ti, ci) * H_b_tot( ti) * U_tot( ti)* u_perp_x / mesh%TriA( ti)
+          ! u_perp < 0: flow is entering this triangle into triangle tj
+          ELSE
+            laddie%divQU( ti) = laddie%divQU( ti) + mesh%TriCw( ti, ci) * H_b_tot( tj) * U_tot( tj)* u_perp_x / mesh%TriA( ti)
+          END IF
+
+          ! V momentum
+          IF (u_perp_y > 0) THEN
+            laddie%divQV( ti) = laddie%divQV( ti) + mesh%TriCw( ti, ci) * H_b_tot( ti) * V_tot( ti)* u_perp_y / mesh%TriA( ti)
+          ELSE
+            laddie%divQV( ti) = laddie%divQV( ti) + mesh%TriCw( ti, ci) * H_b_tot( tj) * V_tot( tj)* u_perp_y / mesh%TriA( ti)
+          END IF
+
+          !IF (par%master) THEN
+          !  WRITE( *, "(I3,I3,I3,I3,A,F15.3,A,F12.6,A,F12.6,A,F18.12)") ti, ci, tj, ei, '  D_x: ', D_x, '  H_b_tot(ti)', H_b_tot( ti), '   U_tot( ti)', U_tot( ti), '   divQU(ti)', laddie%divQU( ti)*mesh%TriA( ti)
+          !END IF     
+        END DO ! DO ci = 1, 3
+
+      END IF ! (laddie%mask_b( ti))
+
+    END DO ! DO ti = mesh%ti1, mesh%ti2
+
+    ! Finalise routine path
+    CALL finalise_routine( routine_name)
+
+  END SUBROUTINE compute_divQUV_upstream
+
+  SUBROUTINE map_laddie_velocities_from_b_to_c_2D( mesh, u_b_partial, v_b_partial, u_c, v_c)
+    ! Calculate velocities on the c-grid for solving the ice thickness equation
+    ! 
+    ! Uses a different scheme then the standard mapping operator, as that one is too diffusive
+        
+    IMPLICIT NONE
+
+    ! In/output variables:
+    TYPE(type_mesh),                        INTENT(IN)    :: mesh
+    REAL(dp), DIMENSION(mesh%ti1:mesh%ti2), INTENT(IN)    :: u_b_partial
+    REAL(dp), DIMENSION(mesh%ti1:mesh%ti2), INTENT(IN)    :: v_b_partial
+    REAL(dp), DIMENSION(mesh%ei1:mesh%ei2), INTENT(OUT)   :: u_c
+    REAL(dp), DIMENSION(mesh%ei1:mesh%ei2), INTENT(OUT)   :: v_c
+      
+    ! Local variables:
+    CHARACTER(LEN=256), PARAMETER                         :: routine_name = 'map_laddie_velocities_from_b_to_c_2D'
+    REAL(dp), DIMENSION(:    ), ALLOCATABLE               :: u_b_tot, v_b_tot
+    INTEGER                                               :: ei, til, tir
+      
+    ! Add routine to path
+    CALL init_routine( routine_name)
+        
+    ! Allocate memory
+    ALLOCATE( u_b_tot( mesh%nTri))
+    ALLOCATE( v_b_tot( mesh%nTri))
+        
+    ! Gather the full b-grid velocity fields to all processes
+    CALL gather_to_all_dp_1D( u_b_partial, u_b_tot)
+    CALL gather_to_all_dp_1D( v_b_partial, v_b_tot)
+
+    ! Map velocities from the b-grid (triangles) to the c-grid (edges)
+    DO ei = mesh%ei1, mesh%ei2
```

Comment:
Is there any difference between this routine and its ice velocity equivalent? Maybe merge them?

Reply:
Yes there is. Also, this is not a final version. I may need several different options as well. So this needs to remain a laddie-specific module